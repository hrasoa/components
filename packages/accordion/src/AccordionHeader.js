// @flow
import * as React from 'react';

type Props = {
  children: string | React.Node,
  controls: string,
  id: string,
  togglePanel: (panelId: string) => void,
  className?: ?string,
  'aria-level'?: string,
  isDisabled?: boolean,
  isExpanded?: boolean,
  isFocused?: boolean,
};

class AccordionHeader extends React.Component<Props> {
  static defaultProps = {
    className: null,
    'aria-level': '3',
    isDisabled: false,
    isExpanded: false,
  }

  componentDidUpdate() {
    if (this.props.isFocused) {
      this.ref.current.focus();
    }
  }

  onTogglePanel = (): void => {
    this.props.togglePanel(this.props.controls);
  }

  onTogglePanel: Function;

  ref: React.Ref<'button'> = React.createRef();

  render() {
    return (
      <dt
        aria-level={this.props['aria-level']}
        role="heading"
      >
        <button
          id={this.props.id}
          className={this.props.className}
          aria-controls={this.props.controls}
          aria-disabled={this.props.isDisabled}
          aria-expanded={this.props.isExpanded}
          type="button"
          ref={this.ref}
          onClick={this.onTogglePanel}
        >
          {this.props.children}
        </button>
      </dt>
    );
  }
}

export default AccordionHeader;
