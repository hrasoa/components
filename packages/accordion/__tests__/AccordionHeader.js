// @flow
import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import AccordionHeader from '../src/AccordionHeader';

const mockTogglePanel = jest.fn();

beforeEach(() => {
  mockTogglePanel.mockClear();
});

describe('AccordionHeader', () => {
  test('initial', () => {
    const component = renderer.create(
      <AccordionHeader
        id="header-1"
        controls="panel-1"
        togglePanel={mockTogglePanel}
      >
        Header 1
      </AccordionHeader>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('is expanded and disabled', () => {
    const component = renderer.create(
      <AccordionHeader
        id="header-1"
        aria-level="2"
        className="accordion__header"
        controls="panel-1"
        isExpanded
        isDisabled
        togglePanel={mockTogglePanel}
      >
        Header 1
      </AccordionHeader>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('focus, expand', () => {
    const wrapper = mount(
      <AccordionHeader
        id="header-1"
        controls="panel-1"
        togglePanel={mockTogglePanel}
      >
        Panel 1
      </AccordionHeader>,
    );
    const ref = wrapper.find('button').first().instance();
    expect(document.activeElement === ref).toBeFalsy();
    wrapper.setProps({ isFocused: true });
    expect(document.activeElement === ref).toBeTruthy();
    wrapper.find('button').simulate('click');
    expect(mockTogglePanel).toHaveBeenCalledTimes(1);
    expect(mockTogglePanel).toHaveBeenLastCalledWith('panel-1');
  });
});
