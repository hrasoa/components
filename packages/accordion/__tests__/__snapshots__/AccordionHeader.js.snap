// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionHeader initial 1`] = `
<dt
  aria-level="3"
  className={null}
  role="heading"
>
  <button
    aria-controls="panel-1"
    aria-disabled={false}
    aria-expanded={false}
    id="header-1"
    onClick={[Function]}
    type="button"
  >
    Header 1
  </button>
</dt>
`;

exports[`AccordionHeader is expanded and disabled 1`] = `
<dt
  aria-level="2"
  className="accordion__header"
  role="heading"
>
  <button
    aria-controls="panel-1"
    aria-disabled={true}
    aria-expanded={true}
    id="header-1"
    onClick={[Function]}
    type="button"
  >
    Header 1
  </button>
</dt>
`;
